### PR TITLE
Skill classification

### PR DESCRIPTION
**Pull request by bexmckinnon**
This pull request covers adding skill categories to questions, implementation of a filtering system for questions and addition of new achievements based on these skill categories.

### CODE REVIEW

File: codewof/programming/management/commands/_QuestionsLoader.py [hunk 1]
```diff
@@ -142,6 +162,13 @@ def load(self):
                     defaults['read_only_lines_top'] = int(question_data.get('number_of_read_only_lines_top', 0))
                     defaults['read_only_lines_bottom'] = int(question_data.get('number_of_read_only_lines_bottom', 0))
 
+                # TODO remove conditional once all difficulty levels assigned
```
Comment (jqi26) [inline codewof/programming/management/commands/_QuestionsLoader.py:165]: Stills needs to be done.

File: codewof/programming/management/commands/_QuestionsLoader.py [hunk 2]
```diff
@@ -151,6 +178,74 @@ def load(self):
...
+                        question.concepts.add(concept)
+                    except ObjectDoesNotExist:
+                        raise KeyNotFoundError(
+                            self.structure_file_path,
+                            concept_slug,
+                            "Concepts"
+                        )
+
+                # Add question contexts
+                context_slugs = question_data.get("contexts", [])
```
Comment (jqi26) [inline codewof/programming/management/commands/_QuestionsLoader.py:216]: Another instance of copy-pasted code due to the similarity of Concepts and Contexts.

**Review by jqi26 — COMMENTED** [review]
This is an interesting project. During my interviews with the CSERG teachers, a couple of them mentioned trying to find certain questions that teach a certain context/want to avoid certain questions. These filters will be really useful for them!

Most of the individual lines of code are good. However, I think that the QuestionContexts and ProgrammingConcepts classes are too similar. This leads to quite a bit of copy+pasted code (e.g. _ProgrammingConceptsLoader.py and _QuestionContextsLoader.py). This will make this code more difficult to maintain in the future (e.g. remembering to update all places if a bug is discovered). I think polymorphism/abstraction would help to reduce a lot of this duplication. However, doing this would require a significant rewrite, and would take a long time.

AFAICT there are no unit tests. These are important to increase confidence to accept the code into the codebase and to help with identifying any faults caused by changes in the future.

There are various todos still in the code. A significant one is that not all questions have been assigned a difficulty yet.

**Comment by JackMorganNZ** [comment]
Closing as #557 built upon this work and is now merged.

**Pull request closed without merging**